src/operations/execute_operation.ts

```diff
@@ -170,60 +165,68 @@ function executeWithServerSelection(
   }
 
   const serverSelectionOptions = { session };
-  function callbackWithRetry(err?: any, result?: any) {
-    if (err == null) {
-      return callback(undefined, result);
-    }
+  function retryOperation(originalError: MongoError, maxWireVersion: number) {
+    const isWriteOperation = operation.hasAspect(Aspect.WRITE_OPERATION);
+    const isReadOperation = operation.hasAspect(Aspect.READ_OPERATION);
 
-    const hasReadAspect = operation.hasAspect(Aspect.READ_OPERATION);
-    const hasWriteAspect = operation.hasAspect(Aspect.WRITE_OPERATION);
-    const itShouldRetryWrite = shouldRetryWrite(err);
+    if (isWriteOperation && !isRetryableWriteError(originalError, maxWireVersion)) {
+      return callback(originalError);
+    }
 
-    if ((hasReadAspect && !isRetryableError(err)) || (hasWriteAspect && !itShouldRetryWrite)) {
-      return callback(err);
+    if (isReadOperation && !isRetryableReadError(originalError)) {
+      return callback(originalError);
     }
 
     if (
-      hasWriteAspect &&
-      itShouldRetryWrite &&
-      err.code === MMAPv1_RETRY_WRITES_ERROR_CODE &&
-      err.errmsg.match(/Transaction numbers/)
+      isWriteOperation &&
+      originalError.code === MMAPv1_RETRY_WRITES_ERROR_CODE &&
+      /Transaction numbers/.test(originalError.errmsg)
     ) {
-      callback(
+      return callback(
```

Comment:
This code happens to not currently be tested - I think we should pull in NODE-2111 as part of this work, since we're trying to ensure spec compliance (it's just one prose test)